src/unix/index_unix.ml

```diff
@@ -85,8 +85,31 @@ module IO : Index.Platform.IO = struct
   let append t buf = append_substring t buf ~off:0 ~len:(String.length buf)
 
   let read t ~off ~len buf =
-    if not t.readonly then assert (t.header ++ off <= t.flushed);
-    Raw.unsafe_read t.raw ~off:(t.header ++ off) ~len buf
+    let off = t.header ++ off in
+    let end_of_value = off ++ Int63.of_int len in
+    if not t.readonly then
+      assert (
+        let total_length = t.flushed ++ Int63.of_int (Buffer.length t.buf) in
+        off <= total_length);
+
+    if t.readonly || end_of_value <= t.flushed then
+      (* Value is entirely on disk *)
+      Raw.unsafe_read t.raw ~off ~len buf
+    else
+      (* Must read some data not yet flushed to disk *)
+      let bytes_from_disk = max 0 (Int63.to_int (t.flushed -- off)) in
+      let bytes_from_buffer = len - bytes_from_disk in
+      let () =
+        if bytes_from_disk > 0 then
+          let read = Raw.unsafe_read t.raw ~off ~len:bytes_from_disk buf in
+          assert (read = bytes_from_disk)
+      in
+      let () =
+        let src_off = max 0 (Int63.to_int (off -- t.flushed)) in
+        let len = min (Buffer.length t.buf - src_off) bytes_from_buffer in
```

Comment:
tricky function :) 
`len` is diferrent from `bytes_from_buffer`  only when we read more than we wrote? maybe we should fail then?

Reply:
> tricky function :)

Yeah, I needed pen and paper for this one :sweat_smile:

> `len` is diferrent from `bytes_from_buffer` only when we read more than we wrote?

The problem here is that a writer may request more bytes than exist (in particular, [here](https://github.com/mirage/index/blob/201c5a66871c41467af0a360dc0ad5d2c28ac3b1/src/index.ml#L664)), in which case we just read until the end of the write buffer (`Buffer.length t.buf - src_off`). I've refactored this a bit to hopefully make it clearer what's going on.

(At some point, we should have a library with buffered reader / writer abstractions so we don't have to re-invent the wheel here.)

Reply:
I see, thanks!